File: im-desktop/src/services/message-sync-service.ts
import { SyncRequest, SyncResponse, SyncCheckpoint } from '../types/message-sync';

const API_BASE = '/api/v1/sync';

function getHeaders(): HeadersInit {
  const userId = localStorage.getItem('userId') || '1';
  return {
    'Content-Type': 'application/json',
    'X-User-Id': userId,
  };
}

export async function pullSync(request: SyncRequest): Promise<SyncResponse> {
  const resp = await fetch(`${API_BASE}/pull`, {
    method: 'POST',
    headers: getHeaders(),
    body: JSON.stringify(request),
  });
  if (!resp.ok) throw new Error(`Sync pull failed: ${resp.status}`);
  return resp.json();
}

export async function fetchHistory(conversationId: number, lastId?: number, limit = 50): Promise<SyncResponse> {
  const params = new URLSearchParams();
  if (lastId) params.set('lastId', String(lastId));
  params.set('limit', String(limit));
  const resp = await fetch(`${API_BASE}/history/${conversationId}?${params}`, {
    headers: getHeaders(),
  });
  if (!resp.ok) throw new Error(`Fetch history failed: ${resp.status}`);
  return resp.json();
}

export async function fetchCheckpoints(deviceId: string): Promise<SyncCheckpoint[]> {
  const resp = await fetch(`${API_BASE}/checkpoints?deviceId=${deviceId}`, {
    headers: getHeaders(),
  });
  if (!resp.ok) throw new Error(`Fetch checkpoints failed: ${resp.status}`);
  const data = await resp.json();
  return data.checkpoints || [];
}

export async function deleteMessageFromHistory(messageId: number): Promise<void> {
  const resp = await fetch(`${API_BASE}/history/${messageId}`, {
    method: 'DELETE',
    headers: getHeaders(),
  });
  if (!resp.ok) throw new Error(`Delete from history failed: ${resp.status}`);
}

export function getDeviceId(): string {
  let deviceId = localStorage.getItem('deviceId');
  if (!deviceId) {
    deviceId = 'device_' + Math.random().toString(36).substring(2, 15);
    localStorage.setItem('deviceId', deviceId);
  }
  return deviceId;
}

File: im-mobile/lib/services/message_sync_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/message_sync.dart';

class MessageSyncService {
  static const String _baseUrl = '/api/v1/sync';

  String get _userId => '1';

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        'X-User-Id': _userId,
      };

  Future<SyncResponse> pullSync(SyncRequest request) async {
    final resp = await http.post(
      Uri.parse('$_baseUrl/pull'),
      headers: _headers,
      body: json.encode(request.toJson()),
    );
    if (resp.statusCode == 200) {
      return SyncResponse.fromJson(json.decode(resp.body));
    }
    throw Exception('Sync pull failed: ${resp.statusCode}');
  }

  Future<SyncResponse> fetchHistory(int conversationId, {int? lastId, int limit = 50}) async {
    final params = <String, String>{'limit': limit.toString()};
    if (lastId != null) params['lastId'] = lastId.toString();
    final uri = Uri.parse('$_baseUrl/history/$conversationId').replace(queryParameters: params);
    final resp = await http.get(uri, headers: _headers);
    if (resp.statusCode == 200) {
      return SyncResponse.fromJson(json.decode(resp.body));
    }
    throw Exception('Fetch history failed: ${resp.statusCode}');
  }

  Future<List<SyncCheckpoint>> fetchCheckpoints(String deviceId) async {
    final resp = await http.get(
      Uri.parse('$_baseUrl/checkpoints?deviceId=$deviceId'),
      headers: _headers,
    );
    if (resp.statusCode == 200) {
      final data = json.decode(resp.body);
      return (data['checkpoints'] as List?)
              ?.map((c) => SyncCheckpoint.fromJson(c))
              .toList() ??
          [];
    }
    throw Exception('Fetch checkpoints failed: ${resp.statusCode}');
  }

  Future<void> deleteMessageFromHistory(int messageId) async {
    final resp = await http.delete(
      Uri.parse('$_baseUrl/history/$messageId'),
      headers: _headers,
    );
    if (resp.statusCode != 200) {
      throw Exception('Delete from history failed: ${resp.statusCode}');
    }
  }
}
